### PR TITLE
Feature/codebreaker

### PR DESCRIPTION
**Pull request by hulkoivan**

### CODE REVIEW

File: lib/codebreaker_gem/autoload.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require 'pry'
+require 'yaml/store'
+require_relative '../codebreaker_gem/version'
```
Comment (n1zx1) [inline lib/codebreaker_gem/autoload.rb]: remove redundant part

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/.bundle/
+/.yardoc
+/_yardoc/
+/coverage/
+/doc/
+/pkg/
+/spec/reports/
+/tmp/
+
```
Comment (n1zx1) [inline .gitignore:10]: add db folder

File: lib/codebreaker_gem/entities/guess.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      MAX_INPUT = 4
+      INPUT_RANGE = (1..6).freeze
+
+      def initialize(digit)
+        @digit = digit
+        validate
+      end
+
+      def parse_code
+        @digit.chars.map { |number| Integer(number) }
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/guess.rb]: @digit.each_char.map(&:to_i)

File: lib/codebreaker_gem/error/error.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+module CodebreakerGem
+  module Error
```
Comment (n1zx1) [inline lib/codebreaker_gem/error/error.rb]: move each class to separate file

File: lib/codebreaker_gem/validator/validator.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+module CodebreakerGem
+  module Validator
+    def validate_guess_input_length(string, max_length, min_length)
+      raise CodebreakerGem::Error::MaxStringLength if string.length > max_length || string.length < min_length
```
Comment (n1zx1) [inline lib/codebreaker_gem/validator/validator.rb:6]: what if there wil be space at the end

File: lib/codebreaker_gem/validator/validator.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    end
+
+    def validate_string_not_empty(string)
+      raise CodebreakerGem::Error::EmptyString if string.empty?
```
Comment (n1zx1) [inline lib/codebreaker_gem/validator/validator.rb]: same as above

File: spec/codebreaker/difficulty_spec.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  it 'check difficulty attempts' do
+    expect(difficulty.attempts).to eq(CodebreakerGem::Entities::Difficulty::DIFFICULTIES[level][:attempts])
+  end
+
+  it 'check difficulty hint' do
+    expect(difficulty.hints).to eq(CodebreakerGem::Entities::Difficulty::DIFFICULTIES[level][:hints])
+  end
+
+  it 'is equal to another difficulty with same attempts and hints' do
```
Comment (n1zx1) [inline spec/codebreaker/difficulty_spec.rb]: remove redutant operator

File: spec/codebreaker/game_spec.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  it 'use hint' do
+    game.use_hint
+    expect(game.user_hints).not_to eq(CodebreakerGem::Entities::Difficulty::DIFFICULTIES[level][:hints])
+  end
+
+  it 'return true if statistic is valid' do
+    attempt = '1234'
+    game.use_attempt(attempt)
+    game.use_hint
+    expect(game.current_statistic).to be_truthy
```
Comment (n1zx1) [inline spec/codebreaker/game_spec.rb]: check each field

File: spec/codebreaker/game_spec.rb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  it 'return true if attempts are equal to 0' do
+    game.instance_variable_set(:@user_attempts, 0)
+    expect(game.lose?).to be_truthy
+  end
+
+  it 'use hint' do
+    game.use_hint
+    expect(game.user_hints).not_to eq(CodebreakerGem::Entities::Difficulty::DIFFICULTIES[level][:hints])
```
Comment (n1zx1) [inline spec/codebreaker/game_spec.rb]: check acuurate values

File: spec/codebreaker/user_code_spec.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+RSpec.describe CodebreakerGem::Entities::UserCode do
+  subject(:code) { described_class.new(secret_code, guess_code) }
+
+  let(:secret_code) { [6, 5, 4, 3] }
+  let(:guess_code) { [2, 6, 6, 6] }
+
+  test_codes = [{ secret_code: [6, 5, 4, 3], guess_code: [5, 6, 4, 3], answer: '++--' },
```
Comment (n1zx1) [inline spec/codebreaker/user_code_spec.rb]: test_data = [
      [6543, 5643, '++--'],
      [6543, 6411, '+-'],
      [6543, 6544, '+++'],
      [6543, 3456, '----'],
      [6543, 6666, '+'],
      [6543, 2666, '-'],
      [6543, 2222, ''],
      [6666, 1661, '++'],
      [1234, 3124, '+---'],
      [1234, 1524, '++-'],
      [1234, 1234, '++++']
    ]

**Review by n1zx1 — CHANGES_REQUESTED** [review]
Add to all tests positive and negative variants

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,6 @@
+---
+language: ruby
```
Comment (n1zx1) [inline .travis.yml]: Do you need traiis config?

File: CODE_OF_CONDUCT.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
+# Contributor Covenant Code of Conduct
```
Comment (n1zx1) [inline CODE_OF_CONDUCT.md]: ?

File: codebracker_gem.gemspec [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
+  spec.require_paths = ['lib']
+
+  spec.add_dependency 'pry'
+  spec.add_development_dependency 'fasterer'
+  spec.add_development_dependency 'rspec'
+  spec.add_development_dependency 'rubocop'
+  spec.add_development_dependency 'rubocop-performance'
+  spec.add_development_dependency 'rubocop-rspec'
+  spec.add_development_dependency 'simplecov'
```
Comment (n1zx1) [inline codebracker_gem.gemspec]: add versions to gems

File: lib/codebreaker_gem.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: false
+
+require 'codebreaker_gem/autoload'
+
+module CodebreakerGem
```
Comment (n1zx1) [inline lib/codebreaker_gem.rb]: do you need empty module here?

File: lib/codebreaker_gem/entities/game.rb [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        @name = CodebreakerGem::Entities::User.new(name.strip)
+        @level = CodebreakerGem::Entities::Difficulty.new(level.downcase.to_sym)
+        @user_attempts = @level.attempts
+        @user_hints = @level.hints
+        @secret_code = CodebreakerGem::Entities::SecretCode.new.generate_secret_code
+        @hint_code = @secret_code.sample(@user_hints)
+      end
+
+      def use_attempt(guess)
+        return if lose?
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/game.rb]: can i use attempt if i already win?

File: lib/codebreaker_gem/entities/secret_code.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module CodebreakerGem
+  module Entities
+    class SecretCode
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/secret_code.rb]: Better name it `SecretCodeGenerator` and method name `call`

File: lib/codebreaker_gem/entities/user.rb [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+      def initialize(name)
+        @name = name
+        validate
+      end
+
+      private
+
+      def validate
+        validate_name_length(name, STRING_MIN_LEGTH, STRING_MAX_LENGTH)
+        validate_input_class(name, String)
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/user.rb:21]: better validate  in this way:
1 -> is it string
2 -> string length

File: lib/codebreaker_gem/entities/user.rb [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+        validate_string_not_empty(name)
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/user.rb]: do you need empty string validation if we already validate string length?

File: lib/codebreaker_gem/entities/user_code.rb [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+      def initialize(secret_code, guess_code)
+        @secret_code = secret_code
+        @guess_code = guess_code
+        @user_result = ''
+      end
+
+      def match
+        guess_code = @guess_code.clone
+        secret_code = @secret_code.clone
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/user_code.rb]: do it when you initialize instance variables

File: lib/codebreaker_gem/entities/user_code.rb [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        strong_match(secret_code, guess_code)
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/user_code.rb]: remove arguments

File: lib/codebreaker_gem/entities/user_code.rb [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+        soft_match(secret_code, guess_code)
```
Comment (n1zx1) [inline lib/codebreaker_gem/entities/user_code.rb]: same as above

File: spec/codebreaker/difficulty_spec.rb [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+
+RSpec.describe CodebreakerGem::Entities::Difficulty do
+  subject(:difficulty) { described_class.new(level) }
+
+  let(:level) { :medium }
+  let(:easy_difficulty) { described_class.new(:easy) }
+  let(:medium_difficulty) { described_class.new(:medium) }
+  let(:hell_difficulty) { described_class.new(:hell) }
+
+  describe '#success tests' do
```
Comment (n1zx1) [inline spec/codebreaker/difficulty_spec.rb]: describe '#success'

File: spec/codebreaker/difficulty_spec.rb [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+    it 'will not raise error if attempts are equal' do
+      expect(difficulty.attempts).to eq(CodebreakerGem::Entities::Difficulty::DIFFICULTIES[level][:attempts])
+    end
+
+    it 'will not raise error if hints are equal' do
+      expect(difficulty.hints).to eq(CodebreakerGem::Entities::Difficulty::DIFFICULTIES[level][:hints])
+    end
+
+    it 'is not equal to another difficulty with less amount attempts and hints' do
```
Comment (n1zx1) [inline spec/codebreaker/difficulty_spec.rb]: bigger than

File: spec/codebreaker/game_spec.rb [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    end
+
+    it 'return true if codes are match' do
+      expect(game.win?).to be_truthy
+    end
+
+    context '#attempts are equal to 0' do
+      before do
+        game.instance_variable_set(:@user_attempts, 0)
+      end
```
Comment (n1zx1) [inline spec/codebreaker/game_spec.rb:30]: empty line

File: spec/codebreaker/game_spec.rb [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+    it 'use attempt' do
+      expect(game.user_attempts).to eq(CodebreakerGem::Entities::Difficulty::DIFFICULTIES[level.to_sym][:attempts] - 1)
+    end
+
+    it 'return true if codes are match' do
+      expect(game.win?).to be_truthy
+    end
+
+    context '#attempts are equal to 0' do
+      before do
```
Comment (n1zx1) [inline spec/codebreaker/game_spec.rb:25]: in one line

File: spec/codebreaker/game_spec.rb [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+    end
+
+    it 'return true if statistic is valid' do
+      expect(game.current_statistic.attempts).to eq(1)
+      expect(game.current_statistic.name).to eq(game.name)
+      expect(game.current_statistic.hints).to eq(1)
+    end
+
+    it 'return true if statistic is saving to db' do
+      expect(game.save_current_statistic).to be_truthy
```
Comment (n1zx1) [inline spec/codebreaker/game_spec.rb]: check if statistics is saving

File: spec/codebreaker/game_spec.rb [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+    it 'return true if users stistic is sorted' do
+      expect(CodebreakerGem::Entities::Game.users_statistic).to be_truthy
+    end
+  end
+
+  describe '#failure tests' do
+    before do
+      game.instance_variable_set(:@guess, attempt)
+      game.instance_variable_set(:@user_attempts, 1)
+    end
```
Comment (n1zx1) [inline spec/codebreaker/game_spec.rb]: empty line

File: spec/codebreaker/user_code_spec.rb [hunk 28]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+RSpec.describe CodebreakerGem::Entities::UserCode do
+  subject(:code) { described_class.new(secret_code, guess_code) }
+
+  describe '#test will pass if all match' do
+    test_codes = [{ secret_code: [6, 5, 4, 3], guess_code: [5, 6, 4, 3], answer: '++--' },
```
Comment (n1zx1) [inline spec/codebreaker/user_code_spec.rb:7]: to let

**Review by n1zx1 — CHANGES_REQUESTED** [review]

**Review by n1zx1 — APPROVED** [review]
